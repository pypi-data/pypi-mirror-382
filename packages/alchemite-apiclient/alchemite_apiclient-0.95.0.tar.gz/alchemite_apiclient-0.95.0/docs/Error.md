# Error

Error response format

## Properties
Name | Type | Description | Notes
------------ | ------------- | ------------- | -------------
**status** | **int** | The HTTP status code associated with this problem. | 
**title** | **str** | A short, human-readable summary of the problem type.  It will not change from occurrence to occurrence of the problem. | 
**detail** | **str** | A human readable explanation specific to this occurrence of the problem. | 
**type** | **str** | An absolute URI that identifies and, when dereferenced, provides human-readable documentation for the problem type.  If no URI is available then the value of this parameter will be \&quot;about:blank\&quot;. | 
**any string name** | **bool, date, datetime, dict, float, int, list, str, none_type** | any string name can be used but the value must be the correct type | [optional]

[[Back to Model list]](../README.md#documentation-for-models) [[Back to API list]](../README.md#documentation-for-api-endpoints) [[Back to README]](../README.md)


