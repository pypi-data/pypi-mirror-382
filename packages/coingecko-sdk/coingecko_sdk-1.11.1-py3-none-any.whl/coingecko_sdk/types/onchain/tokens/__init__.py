# File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

from __future__ import annotations

from .info_recently_updated_get_params import InfoRecentlyUpdatedGetParams as InfoRecentlyUpdatedGetParams
from .info_recently_updated_get_response import InfoRecentlyUpdatedGetResponse as InfoRecentlyUpdatedGetResponse
