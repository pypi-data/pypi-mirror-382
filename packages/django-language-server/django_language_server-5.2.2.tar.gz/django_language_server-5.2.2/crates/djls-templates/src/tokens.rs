use djls_source::Span;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum TagDelimiter {
    Block,
    Variable,
    Comment,
}

impl TagDelimiter {
    pub const CHAR_OPEN: char = '{';
    pub const LENGTH: usize = 2;
    pub const LENGTH_U32: u32 = 2;

    #[must_use]
    pub fn from_input(input: &str) -> Option<Self> {
        let bytes = input.as_bytes();

        if bytes.len() < Self::LENGTH {
            return None;
        }

        if bytes[0] != Self::CHAR_OPEN as u8 {
            return None;
        }

        match bytes[1] {
            b'%' => Some(Self::Block),
            b'{' => Some(Self::Variable),
            b'#' => Some(Self::Comment),
            _ => None,
        }
    }

    #[must_use]
    pub fn opener(self) -> &'static str {
        match self {
            Self::Block => "{%",
            Self::Variable => "{{",
            Self::Comment => "{#",
        }
    }

    #[must_use]
    pub fn closer(self) -> &'static str {
        match self {
            Self::Block => "%}",
            Self::Variable => "}}",
            Self::Comment => "#}",
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Token {
    Block { content: String, span: Span },
    Comment { content: String, span: Span },
    Error { content: String, span: Span },
    Eof,
    Newline { span: Span },
    Text { content: String, span: Span },
    Variable { content: String, span: Span },
    Whitespace { span: Span },
}

impl Token {
    /// Get the content text for content-bearing tokens
    #[must_use]
    pub fn content(&self) -> String {
        match self {
            Token::Block { content, .. }
            | Token::Comment { content, .. }
            | Token::Error { content, .. }
            | Token::Text { content, .. }
            | Token::Variable { content, .. } => content.clone(),
            Token::Whitespace { span, .. } => " ".repeat(span.length_usize()),
            Token::Newline { span, .. } => {
                if span.length() == 2 {
                    "\r\n".to_string()
                } else {
                    "\n".to_string()
                }
            }
            Token::Eof => String::new(),
        }
    }

    /// Get the lexeme as it appears in source
    #[must_use]
    pub fn lexeme(&self) -> String {
        match self {
            Token::Block { content, .. } => format!(
                "{} {} {}",
                TagDelimiter::Block.opener(),
                content,
                TagDelimiter::Block.closer()
            ),
            Token::Variable { content, .. } => format!(
                "{} {} {}",
                TagDelimiter::Variable.opener(),
                content,
                TagDelimiter::Variable.closer()
            ),
            Token::Comment { content, .. } => format!(
                "{} {} {}",
                TagDelimiter::Comment.opener(),
                content,
                TagDelimiter::Comment.closer()
            ),
            Token::Text { content, .. } | Token::Error { content, .. } => content.clone(),
            Token::Whitespace { span, .. } => " ".repeat(span.length_usize()),
            Token::Newline { span, .. } => {
                if span.length() == 2 {
                    "\r\n".to_string()
                } else {
                    "\n".to_string()
                }
            }
            Token::Eof => String::new(),
        }
    }

    #[must_use]
    pub fn offset(&self) -> Option<u32> {
        match self {
            Token::Block { span, .. }
            | Token::Comment { span, .. }
            | Token::Error { span, .. }
            | Token::Variable { span, .. } => {
                Some(span.start().saturating_sub(TagDelimiter::LENGTH_U32))
            }
            Token::Text { span, .. }
            | Token::Whitespace { span, .. }
            | Token::Newline { span, .. } => Some(span.start()),
            Token::Eof => None,
        }
    }

    /// Get the length of the token content
    #[must_use]
    pub fn length(&self) -> u32 {
        let len = match self {
            Token::Block { content, .. }
            | Token::Comment { content, .. }
            | Token::Error { content, .. }
            | Token::Text { content, .. }
            | Token::Variable { content, .. } => content.len(),
            Token::Whitespace { span, .. } | Token::Newline { span, .. } => span.length_usize(),
            Token::Eof => 0,
        };
        u32::try_from(len).unwrap_or(u32::MAX)
    }

    #[must_use]
    pub fn full_span(&self) -> Option<Span> {
        match self {
            Token::Block { span, .. }
            | Token::Comment { span, .. }
            | Token::Variable { span, .. } => {
                Some(span.expand(TagDelimiter::LENGTH_U32, TagDelimiter::LENGTH_U32))
            }
            Token::Error { span, .. } => Some(span.expand(TagDelimiter::LENGTH_U32, 0)),
            Token::Newline { span, .. }
            | Token::Text { span, .. }
            | Token::Whitespace { span, .. } => Some(*span),
            Token::Eof => None,
        }
    }

    #[must_use]
    pub fn content_span(&self) -> Option<Span> {
        match self {
            Token::Block { span, .. }
            | Token::Comment { span, .. }
            | Token::Error { span, .. }
            | Token::Text { span, .. }
            | Token::Variable { span, .. }
            | Token::Whitespace { span, .. }
            | Token::Newline { span, .. } => Some(*span),
            Token::Eof => None,
        }
    }

    #[must_use]
    pub fn full_span_or_fallback(&self) -> Span {
        self.full_span()
            .unwrap_or_else(|| self.content_span_or_fallback())
    }

    #[must_use]
    pub fn content_span_or_fallback(&self) -> Span {
        self.content_span()
            .unwrap_or_else(|| Span::new(self.offset().unwrap_or(0), self.length()))
    }

    #[must_use]
    pub fn spans(&self) -> (Span, Span) {
        let content = self.content_span_or_fallback();
        let full = self.full_span().unwrap_or(content);
        (content, full)
    }
}

#[cfg(test)]
#[derive(Debug, serde::Serialize)]
pub enum TokenSnapshot {
    Block {
        content: String,
        span: (u32, u32),
        full_span: (u32, u32),
    },
    Comment {
        content: String,
        span: (u32, u32),
        full_span: (u32, u32),
    },
    Eof,
    Error {
        content: String,
        span: (u32, u32),
        full_span: (u32, u32),
    },
    Newline {
        span: (u32, u32),
    },
    Text {
        content: String,
        span: (u32, u32),
        full_span: (u32, u32),
    },
    Variable {
        content: String,
        span: (u32, u32),
        full_span: (u32, u32),
    },
    Whitespace {
        span: (u32, u32),
    },
}

#[cfg(test)]
impl Token {
    /// ## Panics
    ///
    /// This may panic on the `full_span` calls, but it's only used in testing,
    /// so it's all good.
    #[must_use]
    pub fn to_snapshot(&self) -> TokenSnapshot {
        match self {
            Token::Block { span, .. } => TokenSnapshot::Block {
                content: self.content(),
                span: span.into(),
                full_span: self.full_span().unwrap().into(),
            },
            Token::Comment { span, .. } => TokenSnapshot::Comment {
                content: self.content(),
                span: span.into(),
                full_span: self.full_span().unwrap().into(),
            },
            Token::Eof => TokenSnapshot::Eof,
            Token::Error { span, .. } => TokenSnapshot::Error {
                content: self.content(),
                span: span.into(),
                full_span: self.full_span().unwrap().into(),
            },
            Token::Newline { span } => TokenSnapshot::Newline { span: span.into() },
            Token::Text { span, .. } => TokenSnapshot::Text {
                content: self.content(),
                span: span.into(),
                full_span: span.into(),
            },
            Token::Variable { span, .. } => TokenSnapshot::Variable {
                content: self.content(),
                span: span.into(),
                full_span: self.full_span().unwrap().into(),
            },
            Token::Whitespace { span } => TokenSnapshot::Whitespace { span: span.into() },
        }
    }
}

#[cfg(test)]
pub struct TokenSnapshotVec(pub Vec<Token>);

#[cfg(test)]
impl TokenSnapshotVec {
    #[must_use]
    pub fn to_snapshot(&self) -> Vec<TokenSnapshot> {
        self.0.iter().map(Token::to_snapshot).collect()
    }
}

#[derive(Debug, Clone)]
pub struct TokenStream(Vec<Token>);

impl TokenStream {
    const CHARS_PER_TOKEN: usize = 6;
    const MIN_CAPACITY: usize = 32;
    const MAX_CAPACITY: usize = 1024;

    #[must_use]
    pub fn with_estimated_capacity(source: &str) -> Self {
        let capacity =
            (source.len() / Self::CHARS_PER_TOKEN).clamp(Self::MIN_CAPACITY, Self::MAX_CAPACITY);
        Self(Vec::with_capacity(capacity))
    }

    #[inline]
    pub fn push(&mut self, token: Token) {
        self.0.push(token);
    }

    /// Get the number of tokens in the stream.
    #[must_use]
    pub fn len(&self) -> usize {
        self.0.len()
    }

    /// Get the number of content tokens (excluding EOF).
    #[must_use]
    pub fn content_len(&self) -> usize {
        self.0.len().saturating_sub(1)
    }

    /// Check if stream is empty.
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
}

impl From<TokenStream> for Vec<Token> {
    fn from(val: TokenStream) -> Self {
        val.0
    }
}

impl IntoIterator for TokenStream {
    type Item = Token;
    type IntoIter = std::vec::IntoIter<Token>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.into_iter()
    }
}
