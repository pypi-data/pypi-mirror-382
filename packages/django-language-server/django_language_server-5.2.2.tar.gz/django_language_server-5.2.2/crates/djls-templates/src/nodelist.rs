use djls_source::Span;

use crate::parser::ParseError;
use crate::tokens::TagDelimiter;

#[salsa::tracked(debug)]
pub struct NodeList<'db> {
    #[tracked]
    #[returns(ref)]
    pub nodelist: Vec<Node>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Node {
    Tag {
        name: String,
        bits: Vec<String>,
        span: Span,
    },
    Comment {
        content: String,
        span: Span,
    },
    Text {
        span: Span,
    },
    Variable {
        var: String,
        filters: Vec<String>,
        span: Span,
    },
    Error {
        span: Span,
        full_span: Span,
        error: ParseError,
    },
}

impl Node {
    #[must_use]
    pub fn span(&self) -> Span {
        match self {
            Node::Tag { span, .. }
            | Node::Variable { span, .. }
            | Node::Comment { span, .. }
            | Node::Text { span, .. }
            | Node::Error { span, .. } => *span,
        }
    }

    #[must_use]
    pub fn full_span(&self) -> Span {
        match self {
            Node::Variable { span, .. } | Node::Comment { span, .. } | Node::Tag { span, .. } => {
                span.expand(TagDelimiter::LENGTH_U32, TagDelimiter::LENGTH_U32)
            }
            Node::Text { span, .. } => *span,
            Node::Error { full_span, .. } => *full_span,
        }
    }

    #[must_use]
    pub fn identifier_span(&self) -> Option<Span> {
        match self {
            Node::Tag { name, span, .. } => {
                // Just the tag name (e.g., "if" in "{% if user.is_authenticated %}")
                Some(span.with_length_usize_saturating(name.len()))
            }
            Node::Variable { var, span, .. } => {
                // Just the variable name (e.g., "user" in "{{ user.name|title }}")
                Some(span.with_length_usize_saturating(var.len()))
            }
            Node::Comment { .. } | Node::Text { .. } | Node::Error { .. } => None,
        }
    }
}
