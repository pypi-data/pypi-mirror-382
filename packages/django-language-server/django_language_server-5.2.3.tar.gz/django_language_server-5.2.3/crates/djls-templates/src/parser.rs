use djls_source::Span;
use serde::Serialize;
use thiserror::Error;

use crate::nodelist::Node;
use crate::tokens::Token;

pub struct Parser {
    tokens: Vec<Token>,
    current: usize,
}

impl Parser {
    #[must_use]
    pub fn new(tokens: Vec<Token>) -> Self {
        Self { tokens, current: 0 }
    }

    pub fn parse(&mut self) -> (Vec<Node>, Vec<ParseError>) {
        let mut nodelist = Vec::with_capacity(self.tokens.len() / 2);
        let mut errors = Vec::with_capacity(4);

        while !self.is_at_end() {
            match self.next_node() {
                Ok(node) => {
                    nodelist.push(node);
                }
                Err(error) => {
                    let (span, full_span) = self
                        .peek_previous()
                        .ok()
                        .or_else(|| self.peek().ok())
                        .map_or(
                            {
                                let empty = Span::new(0, 0);
                                (empty, empty)
                            },
                            super::tokens::Token::spans,
                        );

                    errors.push(error.clone());

                    nodelist.push(Node::Error {
                        span,
                        full_span,
                        error,
                    });

                    if !self.is_at_end() {
                        // Continue parsing even if synchronization fails
                        let _ = self.synchronize();
                    }
                }
            }
        }

        (nodelist, errors)
    }

    fn next_node(&mut self) -> Result<Node, ParseError> {
        let token = self.consume()?;

        match token {
            Token::Block { .. } => self.parse_block(),
            Token::Comment { .. } => self.parse_comment(),
            Token::Eof => Err(ParseError::stream_error(StreamError::AtEnd)),
            Token::Error { .. } => self.parse_error(),
            Token::Newline { .. } | Token::Text { .. } | Token::Whitespace { .. } => {
                self.parse_text()
            }
            Token::Variable { .. } => self.parse_variable(),
        }
    }

    pub fn parse_block(&mut self) -> Result<Node, ParseError> {
        let token = self.peek_previous()?;

        let Token::Block {
            content: content_ref,
            ..
        } = token
        else {
            return Err(ParseError::InvalidSyntax {
                context: "Expected Block token".to_string(),
            });
        };

        let mut parts = content_ref.split_ascii_whitespace();

        let name = parts.next().ok_or(ParseError::EmptyTag)?.to_string();

        let mut bits = Vec::with_capacity(parts.clone().count());
        for part in parts {
            bits.push(part.to_owned());
        }

        let span = token.content_span_or_fallback();

        Ok(Node::Tag { name, bits, span })
    }

    fn parse_comment(&mut self) -> Result<Node, ParseError> {
        let token = self.peek_previous()?;

        let span = token.content_span_or_fallback();
        Ok(Node::Comment {
            content: token.content(),
            span,
        })
    }

    fn parse_error(&mut self) -> Result<Node, ParseError> {
        let token = self.peek_previous()?;

        match token {
            Token::Error { content, span, .. } => {
                let error_text = content.clone();
                let full_span = token.full_span().unwrap_or(*span);
                Err(ParseError::MalformedConstruct {
                    position: full_span.start_usize(),
                    content: error_text,
                })
            }
            _ => Err(ParseError::InvalidSyntax {
                context: "Expected Error token".to_string(),
            }),
        }
    }

    fn parse_text(&mut self) -> Result<Node, ParseError> {
        let first_span = self.peek_previous()?.full_span_or_fallback();
        let start = first_span.start();
        let mut end = first_span.end();

        while let Ok(token) = self.peek() {
            match token {
                Token::Block { .. }
                | Token::Variable { .. }
                | Token::Comment { .. }
                | Token::Error { .. }
                | Token::Eof => break, // Stop at Django constructs, errors, or EOF
                Token::Text { .. } | Token::Whitespace { .. } | Token::Newline { .. } => {
                    // Update end position
                    let token_end = token.full_span_or_fallback().end();
                    end = end.max(token_end);
                    self.consume()?;
                }
            }
        }

        let length = end.saturating_sub(start);
        let span = Span::new(start, length);

        Ok(Node::Text { span })
    }

    fn parse_variable(&mut self) -> Result<Node, ParseError> {
        let token = self.peek_previous()?;

        let Token::Variable {
            content: content_ref,
            ..
        } = token
        else {
            return Err(ParseError::InvalidSyntax {
                context: "Expected Variable token".to_string(),
            });
        };

        let mut parts = content_ref.split('|');

        let var = parts.next().ok_or(ParseError::EmptyTag)?.trim().to_string();

        let filters: Vec<String> = parts.map(|s| s.trim().to_string()).collect();
        let span = token.content_span_or_fallback();

        Ok(Node::Variable { var, filters, span })
    }

    #[inline]
    fn peek(&self) -> Result<&Token, ParseError> {
        self.tokens.get(self.current).ok_or_else(|| {
            if self.tokens.is_empty() {
                ParseError::stream_error(StreamError::Empty)
            } else {
                ParseError::stream_error(StreamError::AtEnd)
            }
        })
    }

    #[inline]
    fn peek_previous(&self) -> Result<&Token, ParseError> {
        if self.current == 0 {
            return Err(ParseError::stream_error(StreamError::BeforeStart));
        }
        self.tokens
            .get(self.current - 1)
            .ok_or_else(|| ParseError::stream_error(StreamError::InvalidAccess))
    }

    #[inline]
    fn is_at_end(&self) -> bool {
        self.current + 1 >= self.tokens.len()
    }

    #[inline]
    fn consume(&mut self) -> Result<&Token, ParseError> {
        if self.is_at_end() {
            return Err(ParseError::stream_error(StreamError::AtEnd));
        }
        self.current += 1;
        self.peek_previous()
    }

    fn synchronize(&mut self) -> Result<(), ParseError> {
        while !self.is_at_end() {
            let current = self.peek()?;
            match current {
                Token::Block { .. }
                | Token::Variable { .. }
                | Token::Comment { .. }
                | Token::Eof => {
                    return Ok(());
                }
                _ => {}
            }
            self.consume()?;
        }
        Ok(())
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
pub enum StreamError {
    AtBeginning,
    BeforeStart,
    AtEnd,
    Empty,
    InvalidAccess,
}

#[derive(Clone, Debug, Error, PartialEq, Eq, Serialize)]
pub enum ParseError {
    #[error("Unexpected token: expected {expected:?}, found {found} at position {position}")]
    UnexpectedToken {
        expected: Vec<String>,
        found: String,
        position: usize,
    },

    #[error("Missing condition in '{tag}' tag at position {position}")]
    MissingCondition { tag: String, position: usize },

    #[error("Missing iterator in 'for' tag at position {position}")]
    MissingIterator { position: usize },

    #[error("Malformed variable at position {position}: {content}")]
    MalformedVariable { position: usize, content: String },

    #[error("Invalid filter syntax at position {position}: {reason}")]
    InvalidFilterSyntax { position: usize, reason: String },

    #[error("Unclosed tag at position {opener}: expected '{expected_closer}'")]
    UnclosedTag {
        opener: usize,
        expected_closer: String,
    },

    #[error("Invalid syntax: {context}")]
    InvalidSyntax { context: String },

    #[error("Empty tag")]
    EmptyTag,

    #[error("Malformed Django construct at position {position}: {content}")]
    MalformedConstruct { position: usize, content: String },

    #[error("Stream error: {kind:?}")]
    StreamError { kind: StreamError },
}

impl ParseError {
    pub fn stream_error(kind: impl Into<StreamError>) -> Self {
        Self::StreamError { kind: kind.into() }
    }
}

#[cfg(test)]
mod tests {
    use serde::Serialize;

    use super::*;
    use crate::lexer::Lexer;

    fn parse_test_template(source: &str) -> Vec<Node> {
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize();
        let mut parser = Parser::new(tokens);
        let (nodes, _errors) = parser.parse();
        nodes
    }

    #[derive(Debug, Clone, PartialEq, Serialize)]
    struct TestNodeList {
        nodelist: Vec<TestNode>,
    }

    #[derive(Debug, Clone, PartialEq, Serialize)]
    #[serde(tag = "type")]
    enum TestNode {
        Tag {
            name: String,
            bits: Vec<String>,
            span: (u32, u32),
            full_span: (u32, u32),
        },
        Comment {
            content: String,
            span: (u32, u32),
            full_span: (u32, u32),
        },
        Text {
            span: (u32, u32),
            full_span: (u32, u32),
        },
        Variable {
            var: String,
            filters: Vec<String>,
            span: (u32, u32),
            full_span: (u32, u32),
        },
        Error {
            span: (u32, u32),
            full_span: (u32, u32),
            error: ParseError,
        },
    }

    impl TestNode {
        fn from_node(node: &Node) -> Self {
            match node {
                Node::Tag { name, bits, span } => TestNode::Tag {
                    name: name.clone(),
                    bits: bits.clone(),
                    span: span.into(),
                    full_span: node.full_span().into(),
                },
                Node::Comment { content, span } => TestNode::Comment {
                    content: content.clone(),
                    span: span.into(),
                    full_span: node.full_span().into(),
                },
                Node::Text { span } => TestNode::Text {
                    span: span.into(),
                    full_span: node.full_span().into(),
                },
                Node::Variable { var, filters, span } => TestNode::Variable {
                    var: var.clone(),
                    filters: filters.clone(),
                    span: span.into(),
                    full_span: node.full_span().into(),
                },
                Node::Error {
                    span,
                    full_span,
                    error,
                } => TestNode::Error {
                    span: span.into(),
                    full_span: full_span.into(),
                    error: error.clone(),
                },
            }
        }
    }

    fn convert_nodelist_for_testing(nodes: &[Node]) -> TestNodeList {
        TestNodeList {
            nodelist: nodes.iter().map(TestNode::from_node).collect(),
        }
    }

    mod html {
        use super::*;

        #[test]
        fn test_parse_html_doctype() {
            let source = "<!DOCTYPE html>";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }

        #[test]
        fn test_parse_html_tag() {
            let source = "<div class=\"container\">Hello</div>";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }

        #[test]
        fn test_parse_html_void() {
            let source = "<input type=\"text\" />";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }
    }

    mod django {
        use super::*;

        #[test]
        fn test_parse_django_variable() {
            let source = "{{ user.name }}";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }

        #[test]
        fn test_parse_django_variable_with_filter() {
            let source = "{{ user.name|title }}";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }

        #[test]
        fn test_parse_filter_chains() {
            let source = "{{ value|default:'nothing'|title|upper }}";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }

        #[test]
        fn test_parse_django_if_block() {
            let source = "{% if user.is_authenticated %}Welcome{% endif %}";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }

        #[test]
        fn test_parse_django_for_block() {
            let source =
                "{% for item in items %}{{ item }}{% empty %}No items{% endfor %}".to_string();
            let nodelist = parse_test_template(&source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }

        #[test]
        fn test_parse_complex_if_elif() {
            let source = "{% if x > 0 %}Positive{% elif x < 0 %}Negative{% else %}Zero{% endif %}";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }

        #[test]
        fn test_parse_django_tag_assignment() {
            let source = "{% url 'view-name' as view %}";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }

        #[test]
        fn test_parse_nested_for_if() {
            let source =
                "{% for item in items %}{% if item.active %}{{ item.name }}{% endif %}{% endfor %}"
                    .to_string();
            let nodelist = parse_test_template(&source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }

        #[test]
        fn test_parse_mixed_content() {
            let source = "Welcome, {% if user.is_authenticated %}
    {{ user.name|title|default:'Guest' }}
    {% for group in user.groups %}
        {% if forloop.first %}({% endif %}
        {{ group.name }}
        {% if not forloop.last %}, {% endif %}
        {% if forloop.last %}){% endif %}
    {% empty %}
        (no groups)
    {% endfor %}
{% else %}
    Guest
{% endif %}!";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }
    }

    mod script {
        use super::*;

        #[test]
        fn test_parse_script() {
            let source = r#"<script type="text/javascript">
    // Single line comment
    const x = 1;
    /* Multi-line
        comment */
    console.log(x);
</script>"#
                .to_string();
            let nodelist = parse_test_template(&source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }
    }

    mod style {
        use super::*;

        #[test]
        fn test_parse_style() {
            let source = r#"<style type="text/css">
    /* Header styles */
    .header {
        color: blue;
    }
</style>"#
                .to_string();
            let nodelist = parse_test_template(&source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }
    }

    mod comments {
        use super::*;

        #[test]
        fn test_parse_comments() {
            let source = "<!-- HTML comment -->{# Django comment #}";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }
    }

    mod whitespace {
        use super::*;

        #[test]
        fn test_parse_with_leading_whitespace() {
            let source = "     hello";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }

        #[test]
        fn test_parse_with_leading_whitespace_newline() {
            let source = "\n     hello";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }

        #[test]
        fn test_parse_with_trailing_whitespace() {
            let source = "hello     ";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }

        #[test]
        fn test_parse_with_trailing_whitespace_newline() {
            let source = "hello     \n";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }
    }

    mod errors {
        use super::*;

        #[test]
        fn test_parse_unclosed_html_tag() {
            let source = "<div>";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }

        #[test]
        fn test_parse_unclosed_django_if() {
            let source = "{% if user.is_authenticated %}Welcome";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }

        #[test]
        fn test_parse_unclosed_django_for() {
            let source = "{% for item in items %}{{ item.name }}";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }

        #[test]
        fn test_parse_unclosed_script() {
            let source = "<script>console.log('test');";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }

        #[test]
        fn test_parse_unclosed_style() {
            let source = "<style>body { color: blue; ";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }

        #[test]
        fn test_parse_unclosed_variable_token() {
            let source = "{{ user";
            let nodelist = parse_test_template(source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }

        // TODO: fix this so we can test against errors returned by parsing
        // #[test]
        // fn test_parse_error_recovery() {
        //     let source = r#"<div class="container">
        //     <h1>Header</h1>
        //     {% %}
        //         {# This if is unclosed which does matter #}
        //         <p>Welcome {{ user.name }}</p>
        //         <div>
        //             {# This div is unclosed which doesn't matter #}
        //         {% for item in items %}
        //             <span>{{ item }}</span>
        //         {% endfor %}
        //     <footer>Page Footer</footer>
        // </div>"#;
        //     let tokens = Lexer::new(source).tokenize().unwrap();
        //     let mut parser = create_test_parser(tokens);
        //     let (nodelist, errors) = parser.parse().unwrap();
        //     let nodelist = convert_nodelist_for_testing(ast.nodelist(parser.db), parser.db);
        //     insta::assert_yaml_snapshot!(nodelist);
        //     assert_eq!(errors.len(), 1);
        //     assert!(matches!(&errors[0], ParserError::EmptyTag));
        // }
    }

    mod full_templates {
        use super::*;

        #[test]
        fn test_parse_full() {
            let source = r#"<!DOCTYPE html>
<html>
    <head>
        <style type="text/css">
            /* Style header */
            .header { color: blue; }
        </style>
        <script type="text/javascript">
            // Init app
            const app = {
                /* Config */
                debug: true
            };
        </script>
    </head>
    <body>
        <!-- Header section -->
        <div class="header" id="main" data-value="123" disabled>
            {% if user.is_authenticated %}
                {# Welcome message #}
                <h1>Welcome, {{ user.name|title|default:'Guest' }}!</h1>
                {% if user.is_staff %}
                    <span>Admin</span>
                {% else %}
                    <span>User</span>
                {% endif %}
            {% endif %}
        </div>
    </body>
</html>"#
                .to_string();
            let nodelist = parse_test_template(&source);
            let test_nodelist = convert_nodelist_for_testing(&nodelist);
            insta::assert_yaml_snapshot!(test_nodelist);
        }
    }
}
