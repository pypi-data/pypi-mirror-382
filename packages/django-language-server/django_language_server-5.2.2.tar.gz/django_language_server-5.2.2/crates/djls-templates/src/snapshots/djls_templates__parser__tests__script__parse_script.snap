---
source: crates/djls-templates/src/parser.rs
expression: test_nodelist
---
nodelist:
  - type: Text
    span:
      - 0
      - 142
    full_span:
      - 0
      - 142
